fix: proposing solution for issue 4

tublian-ai-engineer opened the pull request:
Proposed solution:

## References
- Fixes #4

## Description:

Here is how I approached the problem.

### Problem Summary:

- The ModifyFulfillmentService is supposed to use the EmailService to send email messages through the EmailHttpClient interface when a SalesOrder is modified.
- The EmailHttpClientImpl currently only prints a message instead of sending an actual email.
- The method ModifyFulfillmentService.buildEmailRequest is stubbed and does not fill in the details from the SalesOrder input class.
- The buildEmailRequest method needs to be modified to include details such as the customer email id from the SalesOrder.
- The EmailHttpClientImpl needs to be updated to send an actual email message using SMTP properties.
- Configuration properties for SMTP (host, port, username, password) should be used to facilitate sending emails.


### Solution Plan:

I reviewed multiple solutions and picked the following solution.

The best solution is to integrate the updated ModifyFulfillmentService and EmailHttpClientImpl classes, ensure the application configuration includes the necessary SMTP properties, and test the entire application flow to ensure emails are sent correctly when a SalesOrder is modified.

Here is why I selected the above solution:\n\nThis solution is simple and straightforward, focusing on minimal changes to achieve the desired functionality. It adheres to the principles of YAGNI and KISS by only updating the necessary components and configurations.

### Implementation Plan:

 1. Update `ModifyFulfillmentService` to include the logic for triggering email notifications when a `SalesOrder` is modified.
 2. Update `EmailHttpClientImpl` to handle the email sending logic using the SMTP properties.
 3. Add the necessary SMTP properties to the application configuration file (`application.properties` or `application.yml`).
 4. Write integration tests in a new test file `ModifyFulfillmentServiceIntegrationTest.java` to verify that emails are sent correctly when a `SalesOrder` is modified.
 5. Run the tests to ensure the entire application flow works as expected.

 ### Verification Report:

 One or more check(s) are failing: 
 [ERROR] COMPILATION ERROR :
 [INFO] -------------------------------------------------------------
 [ERROR] /usr/app/repos/tos-oms-webmvc/integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:[6,41] package org.springframework.mail.javamail does not exist
 [ERROR] /usr/app/repos/tos-oms-webmvc/integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:[7,41] package org.springframework.mail.javamail does not exist
 [ERROR] /usr/app/repos/tos-oms-webmvc/integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:[8,41] package org.springframework.mail.javamail does not exist
 [ERROR] /usr/app/repos/tos-oms-webmvc/integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:[10,18] package javax.mail does not exist
 [ERROR] /usr/app/repos/tos-oms-webmvc/integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:[11,27] package javax.mail.internet does not exist
 [ERROR] /usr/app/repos/tos-oms-webmvc/integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:[13,27] package javax.mail.internet does not exist
 [ERROR] /usr/app/repos/tos-oms-webmvc/integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:[21,13] cannot find symbol
   symbol:   class JavaMailSender\n  location: class com.oms.integrations.EmailHttpClientImpl\n[INFO] 7 errors 
   [INFO] -------------------------------------------------------------
   [INFO] ------------------------------------------------------------------------
   [INFO] Reactor Summary for oms 0.0.1-SNAPSHOT:
   [INFO] 
   [INFO] oms ................................................ SUCCESS [  0.069 s]
   [INFO] oms-model .......................................... SUCCESS [  0.489 s]
   [INFO] oms-dal ............................................ SUCCESS [  0.105 s]
   [INFO] oms-gemfire-support ................................ SUCCESS [  0.259 s]
   [INFO] oms-integrations ................................... FAILURE [  0.966 s]
   [INFO] oms-service ........................................ SKIPPED
   [INFO] oms-app ............................................ SKIPPED
   [INFO] ------------------------------------------------------------------------
   [INFO] BUILD FAILURE
   [INFO] ------------------------------------------------------------------------
   [INFO] Total time:  2.007 s
   [INFO] Finished at: 2024-09-03T21:04:58Z
   [INFO] ------------------------------------------------------------------------
   [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project oms-integrations: Compilation failure: Compilation failure: 
   [ERROR] /usr/app/repos/tos-oms-webmvc/integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:[6,41] package org.springframework.mail.javamail does not exist
   [ERROR] /usr/app/repos/tos-oms-webmvc/integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:[7,41] package org.springframework.mail.javamail does not exist
   [ERROR] /usr/app/repos/tos-oms-webmvc/integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:[8,41] package org.springframework.mail.javamail does not exist
   [ERROR] /usr/app/repos/tos-oms-webmvc/integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:[10,18] package javax.mail does not exist
   [ERROR] /usr/app/repos/tos-oms-webmvc/integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:[11,27] package javax.mail.internet does not exist
   [ERROR] /usr/app/repos/tos-oms-webmvc/integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:[13,27] package javax.mail.internet does not exist
   [ERROR] /usr/app/repos/tos-oms-webmvc/integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:[21,13] cannot find symbol
   [ERROR]   symbol:   class JavaMailSender
   [ERROR]   location: class com.oms.integrations.EmailHttpClientImpl
   [ERROR] 
   [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
   [ERROR] Re-run Maven using the -X switch to enable full debug logging.
   [ERROR] 
   [ERROR] For more information about the errors and possible solutions, please read the following articles:
   [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
   [ERROR] 
   [ERROR] After correcting the problems, you can resume the build with the command
   [ERROR]   mvn <args> -rf :oms-integrations


   Please review the changes and let me know how it can be improved.

tublian-ai-engineer committed:
fix: proposing solution for issue 4

eldad-vfunction commented on integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:
```diff
@@ -1,16 +1,52 @@
+            props.put("mail.smtp.starttls.enable", "true");
+            props.put("mail.debug", "true");
+
+            MimeMessage mimeMessage = mailSender.createMimeMessage();
```
Duplicate definition causing a compilation error.

eldad-vfunction commented on integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:
```diff
@@ -1,16 +1,52 @@
-        return String.format("SUCCESS %d", useCnt.incrementAndGet());
+        MimeMessage mimeMessage = javaMailSender.createMimeMessage();
+        try {
+            JavaMailSenderImpl mailSender = (JavaMailSenderImpl) javaMailSender;
```
Why use the impl class?
Why put hard coded default values?

eldad-vfunction commented on integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:
```diff
@@ -1,16 +1,52 @@
+            mailSender.setUsername("your_username");
+            mailSender.setPassword("your_password");
+
+            Properties props = mailSender.getJavaMailProperties();
```
Causing a compilation error - missing import java.util.Properties.

eldad-vfunction commented on integrations/src/main/java/com/oms/dto/EmailRequestDto.java:
```diff
@@ -1,49 +1,16 @@
-    private String emailType;
-
-
-    public EmailRequestDto(String salesOrderNumber, String messageTitle, String messageBody, String emailType) {
```
I don't understand why to remove this constructor - it also causes a compilation error since ModifyFulfillmentService is using it.

eldad-vfunction reviewed (commented):
This is not good enough in my opinion.

nraychaudhuri committed:
Added spring boot starter email dependency

nraychaudhuri commented:
I have added the missing JavaMail dependency. Now, try to fix the errors.

tublian-ai-engineer committed:
fix: proposing solution for issue 4

tublian-ai-engineer commented:
Proposed solution:

## References
- Fixes #4

## Description:

Here is how I approached the problem.

### Problem Summary:

- The `EmailHttpClientImpl` class has a duplicate definition of the `MimeMessage` object, causing a compilation error.
- The use of `JavaMailSenderImpl` class and hard-coded SMTP configuration values in the `sendEmail` method is not appropriate.
- The `EmailHttpClientImpl` class is missing the import statement for `java.util.Properties`, causing a compilation error.
- The constructor for `EmailRequestDto` was removed, causing a compilation error in `ModifyFulfillmentService` which relies on this constructor.


### Solution Plan:

I reviewed multiple solutions and picked the following solution.

This solution addresses all the issues mentioned in the problem summary by making minimal and straightforward changes to the EmailHttpClientImpl.java file. It removes the duplicate MimeMessage object definition, adds the necessary import for java.util.Properties, updates the class to use configuration properties for SMTP settings, and ensures that emails are correctly sent using JavaMailSenderImpl.

Here is why I selected the above solution:

This solution is the best because it adheres to the principles of simplicity and minimalism. It makes the necessary changes without overcomplicating the code, ensuring that the application is flexible, secure, and functional. By focusing on the specific tasks mentioned, it reduces the surface area of the change and avoids unnecessary modifications.


### Implementation Plan:

* 1. Remove the duplicate definition of the MimeMessage object in EmailHttpClientImpl.java.
* 2. Add the import statement for java.util.Properties at the beginning of EmailHttpClientImpl.java.
* 3. Update EmailHttpClientImpl to use configuration properties for SMTP settings instead of hard-coded values.
* 4. Ensure EmailHttpClientImpl correctly sends emails using JavaMailSenderImpl by configuring it with the SMTP properties, including setting the host, port, username, and password.


### Verification Report:

All verification checks are passing


Please review the changes and let me know how it can be improved.

eldad-vfunction commented on service/src/main/java/com/oms/service/ModifyFulfillmentService.java:
```diff
@@ -70,7 +70,7 @@ public OrderLine fetchOrderLineFromItemId(String lineItemId , SalesOrder salesOr
 
     public EmailRequestDto buildEmailRequest(SalesOrder salesOrder) {
-        return new EmailRequestDto("1234","Modify fulfillment","Your line item have been successfully modified for fulfillment","Modify to shipping from store pickup");
+        return new EmailRequestDto(salesOrder.getCustomerOrderId(), "Modify fulfillment", "Your line item has been successfully modified for fulfillment", "Modify to shipping from store pickup");
```
I don't think this matches the parameters of the EmailRequestDto constructor EmailRequestDto(String to, String subject, String body, String from)

eldad-vfunction commented on integrations/src/main/java/com/oms/integrations/EmailHttpClientImpl.java:
```diff
@@ -1,16 +1,62 @@
+            props.put("mail.smtp.auth", "true");
+            props.put("mail.smtp.starttls.enable", "true");
+            props.put("mail.debug", "true");
+
```
There is no need to reconfigure the mail sender every time we send mail - a better option would be to configure it once (say first send)

eldad-vfunction reviewed (commented):
Great improvement.

tublian-ai-engineer committed:
fix: proposing a solution for 4
